trace/opencensus/trace/span_context.py

```diff
@@ -0,0 +1,145 @@
+            from_header=False):
+        if trace_id is None:
+            trace_id = generate_trace_id()
+
+        self.trace_id = self.check_trace_id(trace_id)
+        self.span_id = self.check_span_id(span_id)
+        self.enabled = enabled
+        self.from_header = from_header
+
+    def __str__(self):
+        """Returns a string form of the SpanContext. This is the format of
+        the Trace Context Header and should be forwarded to downstream
+        requests as the X-Cloud-Trace-Context header.
+
+        :rtype: str
+        :returns: String form of the SpanContext.
+        """
+        header = '{}/{};o={}'.format(
+            self.trace_id,
+            self.span_id,
+            int(self.enabled))
+        return header
+
+
+    def check_span_id(self, span_id):
+        """Check the type of span_id to ensure it is int. If it is not int,
+        first try to convert it to int, if failed to convert, then log a
+        warning message and set the span_id to None.
+        
+        :type span_id: int
+        :param span_id: Identifier for the span, unique within a trace.
+        
+        :rtype: int
+        :returns: Span_id for the current span.
+        """
+        if span_id is None:
+            return None
+
+        if span_id == 0:
+            logging.warning(
+                'Span_id {} is invalid, cannot be zero.'.format(span_id))
+            self.from_header = False
+            return None
+
+        if not isinstance(span_id, int):
+            try:
+                span_id = int(span_id)
+            except (TypeError, ValueError):
+                logging.warning(
+                    'The type of span_id should be int, got {}.'.format(
+                        span_id.__class__.__name__))
+                self.from_header = False
+                span_id = None
+
+        return span_id
+
+
+    def check_trace_id(self, trace_id):
+        """Check the format of the trace_id to ensure it is 32-character hex
+        value representing a 128-bit number. Also the trace_id cannot be zero.
+        
+        :type trace_id: str
+        :param trace_id:
+        
+        :rtype: str
+        :returns: Trace_id for the current context.
+        """
+        if trace_id is _INVALID_TRACE_ID:
+            logging.warning(
+                'Trace_id {} is invalid (cannot be all zero), '
+                'generate a new one.'.format(trace_id))
+            self.from_header = False
+            return generate_trace_id()
+
+        trace_id_pattern = re.compile(_TRACE_ID_FORMAT)
+
+        try:
+            match = trace_id_pattern.match(trace_id)
+
+            if match:
+                return trace_id
+            else:
+                logging.warning(
+                    'Trace_id {} does not the match the required format,'
+                    'generate a new one instead.'.format(trace_id))
+                self.from_header = False
+                return generate_trace_id()
+
+        except TypeError:
```

Comment:
I'd prefer that a type be an assertion failure than implicitly creating a new trace id, because it indicates a bug in someone's code.